packages/cicero-core/src/instanceloader.js

```diff
@@ -0,0 +1,205 @@
+        const logicManager = template.getLogicManager();
+        const parserManager = new ParserManager(logicManager.getModelManager(), null, this.instanceKind);
+        const ciceroMarkTransformer = new CiceroMarkTransformer();
+        const ergoEngine = new ErgoEngine();
+        const input = templateMetadata.getSample();
+        const grammar = template.getParserManager().getTemplate();
+
+        // Initialize the parser
+        Util.initParser(
+            parserManager,
+            logicManager,
+            ergoEngine,
+            templateMetadata.getIdentifier(),
+            templateMetadata.getTemplateType(),
+            grammar,
+            templateMetadata.getRuntime(),
+        );
+
+        const data = Util.parseText(parserManager, ciceroMarkTransformer, input, null, null, 'text/sample.md');
+        return InstanceLoader.fromTemplateWithData(Instance, template, data);
+    }
+
+    /**
+     * Create an instance from an archive
+     * @param {*} Instance - the type to construct
+     * @param {Buffer} buffer  - the buffer to a Smart Legal Contract (slc) file
+     * @param {object} options - additional options
+     * @return {Promise<Instance>} a Promise to the instance
+     */
+    static async fromArchive(Instance, buffer, options) {
+        const logicManager = new LogicManager('es6', null, options);
+
+        const method = 'fromArchive';
+        const zip = await JSZip.loadAsync(buffer);
+        const ctoModelFiles = [];
+        const ctoModelFileNames = [];
+
+        // add metadata
+        const packageJsonObject = await InstanceLoader.loadZipFileContents(zip, 'package.json', true, true);
+        // XXX likely we will need to discuss what metadata is relevant / needed for instances
+        // XXX For now hijack the the template metadata
+        const metadata = new InstanceMetadata(
+            packageJsonObject,
+        );
+
+        // add contract data
+        const data = await InstanceLoader.loadZipFileContents(zip, 'data.json', true, true);
+
+        // add contract history
+        const history = await InstanceLoader.loadZipFileContents(zip, 'history.json', true, true);
+
+        //grab instantiator
+        const instantiator = history.length !== 0 ? history[0].currentState.instatiator : null;
+
+        // add template grammar (.md form)
+        const grammar = await InstanceLoader.loadZipFileContents(zip, 'text/grammar.tem.md', false, false);
+
+        // add model files
+        Logger.debug(method, 'Looking for model files');
+        let ctoFiles =  await InstanceLoader.loadZipFilesContents(zip, /model[/\\].*\.cto$/);
+        ctoFiles.forEach(async (file) => {
+            ctoModelFileNames.push(file.name);
+            ctoModelFiles.push(file.contents);
+        });
+        logicManager.getModelManager().addAPModelFiles(ctoModelFiles, ctoModelFileNames, true); // Archives can always be loaded offline
+
+        // load and add the ergo files
```

Comment:
This is looking like a clone of the code to load a template. There must be a better way of sharing that code to ease maintenance?

Reply:
Agreed. As discussed, let's move the contract instance logic into a separate package and put shared code into a `utils` package/class (whichever).